### ISSUE TITLE
Request: toString() for data types?

### ISSUE THREAD

**Issue by Selbi182** [thelinmichael/spotify-web-api-java#226]
Can we get basic toString overrides for data types, such as Album, Track, etc.? It's not a big deal, but it would make debugging so much more convenient.

Example, take a look at this array of PlaylistTracks:
![playlistracks](https://user-images.githubusercontent.com/8850085/93882133-13021980-fce0-11ea-8f81-963b17a227e6.png)
I have no idea which song is which without manually inspecting each element. Eclipse supports ways to manually set detail formatters representations for specific classes:
![grafik](https://user-images.githubusercontent.com/8850085/93882859-121db780-fce1-11ea-93c4-f06ffde26102.png)
But it's rather cumbersome for each and every data type there is.

Are there any reasons we don't at least have some cheap auto-generated toStrings?

**Comment by dargmuesli** [comment]
No, this can surely be added. Please submit a pull request for this! :)

**Comment by Selbi182** [comment]
Alright, give me a few days. Do you have any styling in mind? I'd try to replicate the Lombok otherwise: https://projectlombok.org/features/ToString

**Comment by dargmuesli** [comment]
That sounds like a nice idea. Take your time, thank you :)

**Comment by Selbi182** [comment]
Forget days, this was extremely straight forward: https://github.com/thelinmichael/spotify-web-api-java/pull/227

I've also added an abstract toString() to AbstractModelObject, which enforces every model object subclass to override it, including future ones. You are free to decide if that's okay or not.

**Comment by dargmuesli** [comment]
Hmm, just thinking about it, because you mentioned it... couldn't we just use Lombok? 😅

**Comment by Selbi182** [comment]
I don't like Lombok (anymore). The concept of auto-generated boilerplate is still amazing, but Lombok's biggest downside is the the requirement of a plugin for any IDE working with a project of its kind. There are also some unknown security risks.

Here's an interesting read about the topic: https://artofcode.wordpress.com/2018/12/18/dont-use-lombok/

We use something called immutables.org at work, which works similar to Lombok but uses annotation processing. However, it might be overkill for the Spotify Java wrapper, though the benefits will be drastically reduced boilerplate (free builder() included!).

At the end of the day, it's up to you. I'd gladly help though.

**Comment by dargmuesli** [comment]
I want the codebase to use less boilerplate code anyway, even thought about migrating to Kotlin. But it currently works and there are only rare bug reports or feature requests, so I left it the way it is.
immutables.com looks very interesting! And you have a good point regarding Lombok. Would you mind checking whether we could use immutables.org or whether it is incompatible because of some builder customizations this project uses? I haven't taken a look at most of the code for a long time now.

**Comment by dargmuesli** [comment]
We could also make this part of this year's [Hacktoberfest](https://hacktoberfest.digitalocean.com/) in case you want a slick shirt or let trees be planted 😉

**Comment by Selbi182** [comment]
I've played around with the AlbumSimplified class and Immutables. It seems possible, but it'll be more work than just slapping an annotation on every class. The biggest problem I see right now is the deep-rooted interweaving of hand-written JSON serialization and deserialization. Everything is built around this JsonUtil nested class. Immutables works with Jackson and does most of the nasty work for you... if you set everything up properly.

Bottom line: possible, but definitely a project of lots of mundane labor.

**Comment by dargmuesli** [comment]
> lots of mundane labor

hmm, interestingly, that's something I excel at 💁😂
But not now. As said, I think I'll let this one sit until I got time and leisure. We can use the alternative you've submitted a pull request for for now.
I'll take a closer look at it in about a week or if I can squeeze it in, earlier. I'll try to get it done asap, but I have to take care of other things first. Thank you!

**Comment by dargmuesli** [comment]
@Selbi182, do you want to resubmit the PR so that it counts as [Hacktoberfest](https://hacktoberfest.digitalocean.com/) PR?

**Comment by Selbi182** [comment]
Nah, that's alright. Unless you insist, in which case you can steal it and resubmit it yourself. ;)

**Comment by dargmuesli** [comment]
Nono, I got my four already thanks. Wanted to ask you in case it would help you out :)

**Comment by dargmuesli** [comment]
Thank you, released in https://github.com/thelinmichael/spotify-web-api-java/releases/tag/6.4.0! :)
